src/nvim/tui/tui.c

```diff
@@ -923,19 +925,26 @@ static void tui_mouse_off(UI *ui)
 
 static void tui_set_mode(UI *ui, ModeShape mode)
 {
-  if (!cursor_style_enabled) {
+  if (!cursor_style_enabled || !ui->rgb) {
     return;
   }
   TUIData *data = ui->data;
   cursorentry_T c = data->cursor_shapes[mode];
 
-  if (c.id != 0 && ui->rgb) {
+  if (c.id != 0) {
     int attr = syn_id2attr(c.id);
     if (attr > 0) {
       HlAttrs *aep = syn_cterm_attr2entry(attr);
-      UNIBI_SET_NUM_VAR(data->params[0], aep->rgb_bg_color);
-      unibi_out_ext(ui, data->unibi_ext.set_cursor_color);
+      int attr = ui->rgb ? aep->rgb_ae_attr : aep->cterm_ae_attr;
+      if (attr & HL_INVERSE) {
+        unibi_out_ext(ui, data->unibi_ext.reset_cursor_color);
+      } else {
+        UNIBI_SET_NUM_VAR(data->params[0], aep->rgb_bg_color);
+        unibi_out_ext(ui, data->unibi_ext.set_cursor_color);
+      }
     }
+  } else {
+    unibi_out_ext(ui, data->unibi_ext.reset_cursor_color);
```

Comment:
Should we really forcefully change the cursor when nvim is not capable of restoring the default one ? That might upset users.

Reply:
`\x1b]112\x07` is the escape sequence to restore to the default cursor color specified by the terminal.  I tested with xterm and vte-based terminal and it works fine.  I think this should be the expected behavior. The only problem a user may encounter is that he sets the cursor color using escape sequence before entering the NVIM, but really he should set his cursor color in his terminal emulator.  And he can easily workaround the problem by defining the cursor color in nvim.